File: src/kernel/process/thread.c
/*
 * The BBOS thread is an independent context within a process. All the threads
 * are determined at compile time and can be start and stopped at runtime, but
 * cannot be created or destroyed.
 *
 * Basically we say that thread is subprocess created by a parent process.
 *
 * Copyright (c) ???? Slade Maurer, Alexander Sviridenko
 */

#include <bbos.h>

/*
 * At least one thread should always exist, which means BBOS_NUMBER_OF_THREADS 
 * always > 0. Just a test.
 */
#if BBOS_NUMBER_OF_THREADS > 0

struct bbos_thread bbos_thread_table[BBOS_NUMBER_OF_THREADS];

#endif /* BBOS_NUMBER_OF_THREADS > 0 */

/**
 * bbos_thread_init - Initialize thread.
 */
bbos_return_t
bbos_thread_init(bbos_thread_id_t id)
{
  bbos_thread_table[id].id = id;
}

/**
 * bbos_thread_suspend - Suspend thread.
 * @tid: Thread identifier.
 *
 * Return value:
 *
 * Generic error code.
 */
bbos_return_t
bbos_thread_suspend(bbos_thread_id_t tid)
{
  assert(tid <= BBOS_NUMBER_OF_THREADS);

  return bbos_scheduler_suspend_thread(tid);
}

/**
 * bbos_thread_resume - Resume thread activity.
 * @tid: Thread identifier.
 *
 * Return value:
 *
 * Generic error code.
 */
bbos_return_t
bbos_thread_resume(bbos_thread_id_t tid)
{
  assert(tid <= BBOS_NUMBER_OF_THREADS);

  return bbos_scheduler_resume_thread(tid);
}

bbos_return_t
bbos_thread_destroy(bbos_thread_id_t tid)
{
  return bbos_scheduler_remove_thread(tid);
}

File: bbos/kernel/thread.py

__copyright__ = "Copyright (c) 2011 Slade Maurer, Alexander Sviridenko"

from bbos.component import Component

class Thread(Component):
    """
    Thread base class.
    """
    def __init__(self, name, entry, alias=None):
        Component.__init__(self, name)
        self.entry = entry
        if not alias:
            alias = entry
        self.alias = alias

    def set_alias(self, alias):
        pass

    def get_entry(self):
        """
        Get thread entry function.
        """
        return self.entry

    def get_alias(self):
        """
        Get alias of entry function.
        """
        return self.alias

